### DISCUSSION TITLE
Any way of obtaining dimensions of a structure piece?

### DISCUSSION THREAD

**Discussion by Daniel366Cobra** [FabricMC/fabric-api#1968] [Mod Dev Support] [answered]
Currently working on a processor that should randomly mossify N lower blocks and erode N upper blocks of a structure feature.
However, I don't seem to be able to find a way to obtain the dimensions (in my case, height) of a structure. 
I talked to modders and they suggest that Forge has an overload of the **process** method which actually passes the structure's **Template** and that info on dimensions could be obtained that way, but Fabric does not seem to have that overload available.

So... is there any way of getting the structure's dimensions or template in Fabric?
Thank you in advance.

**Comment by warjort** [comment, marked as answer]
I am no expert on structures.

But maybe the structure post processor is what you are looking for?
It gets passed the list of structure pieces from which you can get the bounding box for the structure.

See the ````WoodlandMansionFeature::postPlace```` for example.

**Reply by Daniel366Cobra** [reply]
`postPlace` is definitely a way to do that, but that's slightly different from what I am trying to achieve.
My aim is to make a StructureProcessor which could be applied to any Structure to mossify cobblestones and erode stone bricks.

Forge `StructureProcessor` has this (referenced from Forge javadocs):
`process(IWorldReader p_230386_1_,
              BlockPos p_230386_2_, BlockPos p_230386_3_,
              Template.BlockInfo p_230386_4_, Template.BlockInfo p_230386_5_,
              PlacementSettings p_230386_6_,
             @Nullable Template template)`

Forge modders told me the last `@Nullable` argument could be used to extract structure dimensions from a template it passes.
Fabric does not offer this overload for a `process` method unfortunately. That is why I am asking people here.
Sure, if there is absolutely no other way of doing this, hard-coding the postprocessing would be the only solution.
Thank you for your answer anyway.

**Reply by warjort** [reply]
Yes, in yarn terms forge rewrites calls to the static method ````Structure.process()```` to pass the Structure as an additional parameter.
That static method passes the new parameter onto a new overloaded ````StructureProcessor.process()```` also with the additional parameter.

You would do the same thing in fabric with mixins.

Doing it exactly the same way as forge would require using redirect mixins that overwrites the method invocations with different calls that have the new parameter.
However, this might interfere with other mods trying to do the same or similar things.

A more conservative approach would be to change ````Structure.place()```` to set and unset a thread local that contains the structure.
You could then access the thread local structure reference in your ````process()```` method.

**Reply by Daniel366Cobra** [reply]
Thank you for your help!
